File: packages/dashcam_player/android/src/main/kotlin/com/dashcam/player/DashcamConfig.kt
package com.dashcam.player

/**
 * Configuration for dashcam RTSP streaming.
 *
 * Supports developer overrides with F9 dashcam defaults as fallback.
 * All endpoints can be overridden individually — if not provided,
 * they're built from F9 default paths + the effective base URL.
 */
class DashcamConfig(map: Map<String, Any>? = null) {

    companion object {
        // F9 Dashcam defaults
        const val DEFAULT_IP = "192.168.169.1"
        const val DEFAULT_RTSP_PORT = 554
        const val DEFAULT_HTTP_PORT = 80
        const val DEFAULT_USER_AGENT = "HiCamera"

        // F9 default endpoint paths
        const val DEFAULT_HEARTBEAT_PATH = "/app/getparamvalue?param=rec"
        const val DEFAULT_ENTER_RECORDER_PATH = "/app/enterrecorder"
        const val DEFAULT_GET_MEDIA_INFO_PATH = "/app/getmediainfo"
        const val DEFAULT_START_LIVE_PATH = "/?custom=1&cmd=2015&par="
        const val DEFAULT_SWITCH_CAMERA_PATH = "/app/setparamvalue?param=switchcam&value="

        const val CAMERA_FRONT = 0
        const val CAMERA_REAR = 1
        const val CAMERA_PIP = 2
    }

    // Effective values (override or F9 default)
    val ip: String = (map?.get("ip") as? String) ?: DEFAULT_IP
    val rtspPort: Int = (map?.get("rtspPort") as? Number)?.toInt() ?: DEFAULT_RTSP_PORT
    val httpPort: Int = (map?.get("httpPort") as? Number)?.toInt() ?: DEFAULT_HTTP_PORT
    val userAgent: String = (map?.get("userAgent") as? String) ?: DEFAULT_USER_AGENT

    // Base URLs
    private val httpBase: String = "http://$ip:$httpPort"

    // RTSP URL (full override or built from ip + port)
    val rtspUrl: String = (map?.get("rtspUrl") as? String) ?: "rtsp://$ip:$rtspPort/"

    // HTTP API endpoints (full override or built from base + F9 default path)
    val apiHeartbeat: String =
        (map?.get("heartbeatEndpoint") as? String) ?: "$httpBase$DEFAULT_HEARTBEAT_PATH"
    val apiEnterRecorder: String =
        (map?.get("enterRecorderEndpoint") as? String) ?: "$httpBase$DEFAULT_ENTER_RECORDER_PATH"
    val apiGetMediaInfo: String =
        (map?.get("getMediaInfoEndpoint") as? String) ?: "$httpBase$DEFAULT_GET_MEDIA_INFO_PATH"
    val apiStartLive: String =
        (map?.get("startLiveEndpoint") as? String) ?: "$httpBase$DEFAULT_START_LIVE_PATH"
    val apiSwitchCamera: String =
        (map?.get("switchCameraEndpoint") as? String) ?: "$httpBase$DEFAULT_SWITCH_CAMERA_PATH"
}

File: packages/dashcam_player/lib/dashcam_config.dart
/// Configuration for the dashcam player plugin.
///
/// All fields are optional. When not specified, F9 dashcam defaults are used.
/// Override any combination of network, endpoint, or RTSP settings for your dashcam.
///
/// ```dart
/// // F9 dashcam defaults (no config needed)
/// DashcamPlayerController()
///
/// // Custom dashcam — just change IP and ports
/// DashcamPlayerController(
///   config: DashcamConfig(ip: '192.168.0.100', rtspPort: 8554),
/// )
///
/// // Different dashcam brand — custom endpoints too
/// DashcamPlayerController(
///   config: DashcamConfig(
///     ip: '10.0.0.1',
///     heartbeatEndpoint: '/cgi-bin/ping',
///     enterRecorderEndpoint: '/cgi-bin/start_rec',
///     startLiveEndpoint: '/cgi-bin/live?cam=',
///     switchCameraEndpoint: '/cgi-bin/switch?ch=',
///   ),
/// )
/// ```
class DashcamConfig {
  // ── F9 Dashcam defaults ──────────────────────────────────────

  // Network
  static const String _defaultIp = '192.168.169.1';
  static const int _defaultRtspPort = 554;
  static const int _defaultHttpPort = 80;
  static const String _defaultUserAgent = 'HiCamera';

  // Endpoint paths (appended to http://ip:httpPort)
  static const String _defaultHeartbeatPath = '/app/getparamvalue?param=rec';
  static const String _defaultEnterRecorderPath = '/app/enterrecorder';
  static const String _defaultGetMediaInfoPath = '/app/getmediainfo';
  static const String _defaultStartLivePath = '/?custom=1&cmd=2015&par=';
  static const String _defaultSwitchCameraPath =
      '/app/setparamvalue?param=switchcam&value=';

  // ── Configurable fields (all optional) ───────────────────────

  /// Dashcam IP address. Default: `192.168.169.1`
  final String? ip;

  /// RTSP port. Default: `554`
  final int? rtspPort;

  /// HTTP port. Default: `80`
  final int? httpPort;

  /// User-Agent header for HTTP requests. Default: `HiCamera`
  final String? userAgent;

  /// Full heartbeat URL override.
  /// Default: `http://{ip}:{httpPort}/app/getparamvalue?param=rec`
  final String? heartbeatEndpoint;

  /// Full enter-recorder URL override.
  /// Default: `http://{ip}:{httpPort}/app/enterrecorder`
  final String? enterRecorderEndpoint;

  /// Full get-media-info URL override.
  /// Default: `http://{ip}:{httpPort}/app/getmediainfo`
  final String? getMediaInfoEndpoint;

  /// Full start-live-preview URL override. Camera index is appended.
  /// Default: `http://{ip}:{httpPort}/?custom=1&cmd=2015&par=`
  final String? startLiveEndpoint;

  /// Full switch-camera URL override. Camera index is appended.
  /// Default: `http://{ip}:{httpPort}/app/setparamvalue?param=switchcam&value=`
  final String? switchCameraEndpoint;

  /// Full RTSP URL override.
  /// Default: `rtsp://{ip}:{rtspPort}/`
  final String? rtspUrl;

  const DashcamConfig({
    this.ip,
    this.rtspPort,
    this.httpPort,
    this.userAgent,
    this.heartbeatEndpoint,
    this.enterRecorderEndpoint,
    this.getMediaInfoEndpoint,
    this.startLiveEndpoint,
    this.switchCameraEndpoint,
    this.rtspUrl,
  });

  // ── Effective values (override or F9 default) ────────────────

  String get effectiveIp => ip ?? _defaultIp;
  int get effectiveRtspPort => rtspPort ?? _defaultRtspPort;
  int get effectiveHttpPort => httpPort ?? _defaultHttpPort;
  String get effectiveUserAgent => userAgent ?? _defaultUserAgent;

  String get _httpBase => 'http://$effectiveIp:$effectiveHttpPort';

  /// Heartbeat URL
  String get effectiveHeartbeatEndpoint =>
      heartbeatEndpoint ?? '$_httpBase$_defaultHeartbeatPath';

  /// Enter recorder mode URL
  String get effectiveEnterRecorderEndpoint =>
      enterRecorderEndpoint ?? '$_httpBase$_defaultEnterRecorderPath';

  /// Get media info URL
  String get effectiveGetMediaInfoEndpoint =>
      getMediaInfoEndpoint ?? '$_httpBase$_defaultGetMediaInfoPath';

  /// Start live preview URL (camera index appended at runtime)
  String get effectiveStartLiveEndpoint =>
      startLiveEndpoint ?? '$_httpBase$_defaultStartLivePath';

  /// Switch camera URL (camera index appended at runtime)
  String get effectiveSwitchCameraEndpoint =>
      switchCameraEndpoint ?? '$_httpBase$_defaultSwitchCameraPath';

  /// RTSP URL
  String get effectiveRtspUrl =>
      rtspUrl ?? 'rtsp://$effectiveIp:$effectiveRtspPort/';

  /// Serialize only non-null values for passing to native side
  Map<String, dynamic> toMap() {
    final map = <String, dynamic>{};
    if (ip != null) map['ip'] = ip;
    if (rtspPort != null) map['rtspPort'] = rtspPort;
    if (httpPort != null) map['httpPort'] = httpPort;
    if (userAgent != null) map['userAgent'] = userAgent;
    if (heartbeatEndpoint != null) map['heartbeatEndpoint'] = heartbeatEndpoint;
    if (enterRecorderEndpoint != null) map['enterRecorderEndpoint'] = enterRecorderEndpoint;
    if (getMediaInfoEndpoint != null) map['getMediaInfoEndpoint'] = getMediaInfoEndpoint;
    if (startLiveEndpoint != null) map['startLiveEndpoint'] = startLiveEndpoint;
    if (switchCameraEndpoint != null) map['switchCameraEndpoint'] = switchCameraEndpoint;
    if (rtspUrl != null) map['rtspUrl'] = rtspUrl;
    return map;
  }
}
